node/src/store/header_ranges.rs

```diff
@@ -0,0 +1,455 @@
+    fn len(&self) -> u64 {
+        self.end().saturating_sub(*self.start()) + 1
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Default, Serialize)]
+pub struct HeaderRanges(SmallVec<[HeaderRange; 2]>);
+
+#[derive(Debug, Error)]
+pub enum HeaderRangeError {
+    #[error("Overlapping range: {0}, {1}")]
+    RangeOverlap(u64, u64),
+}
+
+pub(crate) trait HeaderRangesExt {
+    /// check whether sub-ranges do not overlap
+    fn validate(&self) -> Result<(), HeaderRangeError>;
+    /// Check whether provided `to_insert` range can be inserted into the header ranges represented
+    /// by self. New range can be inserted ahead of all existing ranges to allow syncing from the
+    /// head but otherwise, only growing the existing ranges is allowed.
+    /// Returned [`RangeScanResult`] contains information necessary to persist the range
+    /// modification in the database manually, or one can call [`update_range`] to modify ranges in
+    /// memory.
+    fn check_range_insert(&self, to_insert: &HeaderRange) -> Result<RangeScanResult, StoreError>;
+    /// Modify the header ranges, committing insert previously checked with [`check_range_insert`]
+    fn update_range(&mut self, scan_information: RangeScanResult);
+    /// return whether range is empty
+    fn is_empty(&self) -> bool;
+    /// Return highest height in the range
+    fn head(&self) -> Option<u64>;
+}
+
+impl HeaderRangesExt for HeaderRanges {
+    fn validate(&self) -> Result<(), HeaderRangeError> {
+        let mut prev: Option<&HeaderRange> = None;
+        for current in &self.0 {
+            if let Some(prev) = prev {
+                if current.start() > prev.end() {
+                    return Err(HeaderRangeError::RangeOverlap(
+                        *current.start(),
+                        *prev.end(),
+                    ));
+                }
+            }
+            prev = Some(current);
+        }
+        Ok(())
+    }
+
+    /// Commit previously calculated `check_range_insert`
+    fn update_range(&mut self, scan_information: RangeScanResult) {
+        let RangeScanResult {
+            range_index,
+            range,
+            range_to_remove,
+        } = scan_information;
+
+        if self.0.len() == range_index {
+            self.0.push(range);
+        } else {
+            self.0[range_index] = range;
+        }
+
+        if let Some(to_remove) = range_to_remove {
+            self.0.remove(to_remove);
+        }
+    }
+
+    fn check_range_insert(&self, to_insert: &HeaderRange) -> Result<RangeScanResult, StoreError> {
+        let Some(head_range) = self.0.last() else {
+            // Empty store case
+            return Ok(RangeScanResult {
+                range_index: 0,
+                range: to_insert.clone(),
+                range_to_remove: None,
+            });
+        };
+
+        // allow inserting a new header range in front of the current head range
+        if *to_insert.start() > head_range.end() + 1 {
```

Comment:
This is a bit confusing. It took me some time to understand the purpose of `+ 1`. As I understand if the last stored range is `10..=11` you want to be able to add `13..=14` but not `12..=14`? Is it because if they are continues you want to merge them into `10..=14`?.

I think you need to add in the comment that a gap needs to be present.

Reply:
This has no impact on which inserts are valid, but is mearly here to make sure that in case the ranges are contiguous, they merge into one and we don't create slightly pathological ranges like [1..=2, 3..=5], since they have simpler representation.
Added comment about it.